### PR TITLE
feat(engine): support bpmn converging inclusive gateway

### LINKED ISSUE #10031: Support BPMN Converging Inclusive Gateway
**Is your feature request related to a problem? Please describe.**
The BPMN symbol for the inclusive gateway is supported with
- #9747 

However, only the diverging behavior was added. Users may want to use the converging behavior as well.

**Describe the solution you'd like**
- remove the validation that restricts the number of incoming sequence flows to an inclusive gateway to max 1.
- only activate the inclusive gateway when (see discussion https://github.com/camunda/zeebe/pull/9747#discussion_r925770530):
  - there are no active children of its flow scope instance (we can check this already),
  - or if all incoming sequence flows were taken at least once (we can check this already),
  - or if no path can be found from any of the active children to the inclusive gateway, excluding paths that end on an already taken incoming sequence flow of the gateway. (but we can't check this yet).

**Describe alternatives you've considered**
Users can use a combination of parallel and exclusive gateways to build similar joining logic.

**Additional context**
Spec: https://www.omg.org/spec/BPMN/2.0.2/PDF

Product-Hub:
- https://github.com/camunda/product-hub/issues/364

_EDIT: Update the solution about active children based on the BPMN spec_

### PR DESCRIPTION
**Pull request by lzgabel**
## Description
This PR adds support for converging/joining inclusive gateway.

## Related issues
closes #10031

## Definition of Done
Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by lzgabel** [comment]
:wave: Hi @korthout. Please take a moment to review this feature. 🙇

**Comment by Zelldon** [comment]
@camunda/zeebe-process-automation has someone time to review it? \cc @megglos

**Comment by saig0** [comment]
@lzgabel thank you for your contribution. This is a valuable addition to the Zeebe engine. :tada: 

I'll review the changes soon. :eyes:

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java [hunk 1]
```diff
@@ -187,4 +200,149 @@ private Either<String, ElementInstance> hasNonInterruptedFlowScope(
...
+      final BpmnElementContext context,
+      final ExecutableFlowElement executableFlowElement,
+      final ExecutableProcess process) {
+    final var targetElementId = executableFlowElement.getId();
+    final var activateElementIds = findActivateElementInFlowScope(context, targetElementId);
+    final Optional<DirectBuffer> exists =
+        activateElementIds.stream()
+            .filter(
+                elementId -> doesPathExist(process, elementId, targetElementId, new HashSet<>()))
+            .findAny();
```
Comment (saig0) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:252]: :wrench: We should exclude paths ending on an incoming sequence flow of the gateway that was already taken. All incoming flows need to be taken only once.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java [hunk 2]
```diff
@@ -187,4 +200,149 @@ private Either<String, ElementInstance> hasNonInterruptedFlowScope(
...
+                flowId -> {
+                  final DirectBuffer elementId =
+                      process
+                          .getElementById(flowId, ExecutableSequenceFlow.class)
+                          .getTarget()
+                          .getId();
+
+                  return doesPathExist(process, elementId, targetElementId, new HashSet<>());
+                })
+            .collect(Collectors.toSet());
```
Comment (saig0) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:294]: :wrench: We should exclude paths ending on an incoming sequence flow of the gateway that was already taken. All incoming flows need to be taken only once.

**Review by saig0 — CHANGES_REQUESTED** [review]
@lzgabel great work on a very complex topic. :medal_sports: 

I added comments to improve the implementation and add missing pieces. Please have a look. :cake:

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java [hunk 3]
```diff
@@ -187,4 +200,149 @@ private Either<String, ElementInstance> hasNonInterruptedFlowScope(
...
+          sourceElement.getOutgoing().stream()
+              .map(ExecutableSequenceFlow::getTarget)
+              .map(ExecutableFlowNode::getId)
+              .collect(Collectors.toList());
+
+      if (sourceElement instanceof ExecutableActivity) {
+        final List<ExecutableBoundaryEvent> boundaryEvents =
+            ((ExecutableActivity) sourceElement).getBoundaryEvents();
+        boundaryEvents.forEach(event -> sourceElementIds.add(event.getId()));
+      }
```
[the remark this answers is not in the document]
Comment (lzgabel) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:341]: > If the element is inside a subprocess then the subprocess could have a boundary event that can reach the inclusive gateway.

Hi @saig0 . Sorry for the delay, I don't really understand this scenario. We try to get whether there is an activated element in the same scope as the inclusive gateway, and use the activated element ID to determine whether there is a path to reach the inclusive gateway. Even if there is an activated element in the sub-process, when getting activated elements in the same scope as the inclusive gateway, only the sub-process will be obtained

[the remark this answers is not in the document]
Comment (saig0) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:341]: True. You're right. :+1: Please ignore my comment. :sweat_smile:

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java [hunk 4]
```diff
@@ -76,7 +86,32 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
...
+                    BufferUtil.bufferAsString(sequenceFlowId)));
+  }
+
+  private int cleanupSequenceFlowsTaken(final ProcessInstanceRecord value) {
+    // delete all pending sequence flows to this element
+    cleanupPendingSequenceFlows(value);
+
+    final int numberOfTakenSequenceFlows =
+        elementInstanceState.getAllNumberOfTakenSequenceFlows(
+            value.getFlowScopeKey(), value.getElementIdBuffer());
```
[the remark this answers is not in the document]
Comment (lzgabel) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:113]: Hi @saig0. Sorry for the delay.  Can this question be answered through previous comments?

https://github.com/camunda/camunda/issues/10031#issuecomment-1966241232

[the remark this answers is not in the document]
Comment (saig0) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:113]: > On ACTIVATING, delete all pending sequence flows to the gateway

After reading the BPMN spec again, I think this statement is not fully correct. :sweat_smile: 

I described an interesting edge case in this comment https://github.com/camunda/camunda/issues/10031#issuecomment-2132641934.

However, I recommend ignoring the edge case for this PR and focusing on the primary use case. So, let's keep the logic as it is. :+1:

[the remark this answers is not in the document]
Comment (lzgabel) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java]: 👋 Hi @saig0. Let inclusive gateway consume one token per incoming sequence flow like a parallel gateway using `getNumberOfTakenSequenceFlows` method when activate inclusive gateway.

**Comment by lzgabel** [comment]
👋 Hi @saig0. Sorry for the delay. Please take a moment to review the latest commit when you have time. 🙇

After this PR review is finished, I'll resolve the conflicts and squash all commits and rebase it on `main`.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java [hunk 5]
```diff
@@ -709,4 +714,541 @@ public void shouldEvaluateConditionWithoutDefaultFlow() {
...
+    processBuilder
+        .startEvent()
+        .parallelGateway("fork")
+        .inclusiveGateway("join")
+        .endEvent()
+        .moveToNode("fork")
+        .intermediateThrowEvent("throw", b -> b.link("link"));
+
+    final BpmnModelInstance process =
+        processBuilder.linkCatchEvent("catch").link("link").connectTo("join").done();
```
Comment (saig0) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java:1194]: :wrench: We could use the fluent builder to add the link catch event by using `.moveToProcess()`.

Let's extend the process by a service task before the link throw event to verify that the process instance waits until the second token arrives at the inclusive gateway. 

```suggestion
    final var process =
        Bpmn.createExecutableProcess(PROCESS_ID)
            .startEvent()
            .parallelGateway("fork")
            .task("task-1")
            .inclusiveGateway("join")
            .endEvent()
            .moveToNode("fork")
            .serviceTask("task-2", s -> s.zeebeJobType("test"))
            .intermediateThrowEvent("throw", b -> b.link("link"))
            .moveToProcess(PROCESS_ID)
            .linkCatchEvent("catch")
            .link("link")
            .connectTo("join")
            .done();
```

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java [hunk 6]
```diff
@@ -709,4 +714,541 @@ public void shouldEvaluateConditionWithoutDefaultFlow() {
...
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .withElementId("join")
+                .withIntent(ProcessInstanceIntent.ELEMENT_COMPLETED))
+        .extracting(e -> e.getValue().getElementId(), Record::getIntent)
+        .containsOnlyOnce(tuple("join", ProcessInstanceIntent.ELEMENT_COMPLETED));
```
Comment (saig0) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java:1209]: :wrench: We should extend the verification to ensure that the process instance waits at the inclusive gateway. 

```suggestion
    assertThat(
            RecordingExporter.processInstanceRecords()
                .withProcessInstanceKey(processInstanceKey)
                .limitToProcessInstanceCompleted())
        .extracting(e -> e.getValue().getElementId(), Record::getIntent)
        .containsSubsequence(
            tuple("task-1", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
        .containsSubsequence(
            tuple("task-2", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
        .containsOnlyOnce(
            tuple("join", ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple(PROCESS_ID, ProcessInstanceIntent.ELEMENT_COMPLETED));
```

Comment (saig0) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java:1209]: :x: The test case fails with the updated process and verifications. 

```
Expecting actual:
  [("process", ACTIVATE_ELEMENT),
    ("process", ELEMENT_ACTIVATING),
    ("process", ELEMENT_ACTIVATED),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ACTIVATE_ELEMENT),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_ACTIVATING),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_ACTIVATED),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", COMPLETE_ELEMENT),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_COMPLETING),
    ("startEvent_41f087b3-4a7a-49cc-820d-edc51b866932", ELEMENT_COMPLETED),
    ("sequenceFlow_6d79d12e-9519-4063-8c46-c72d2f192522", SEQUENCE_FLOW_TAKEN),
    ("fork", ACTIVATE_ELEMENT),
    ("fork", ELEMENT_ACTIVATING),
    ("fork", ELEMENT_ACTIVATED),
    ("fork", ELEMENT_COMPLETING),
    ("fork", ELEMENT_COMPLETED),
    ("sequenceFlow_5ce126ed-a8d1-4ba4-a4c1-d6295507714b", SEQUENCE_FLOW_TAKEN),
    ("task-2", ACTIVATE_ELEMENT),
    ("sequenceFlow_f1a03bab-2312-461b-acaa-a5872194f8f9", SEQUENCE_FLOW_TAKEN),
    ("task-1", ACTIVATE_ELEMENT),
    ("task-2", ELEMENT_ACTIVATING),
    ("task-2", ELEMENT_ACTIVATED),
    ("task-1", ELEMENT_ACTIVATING),
    ("task-1", ELEMENT_ACTIVATED),
    ("task-1", COMPLETE_ELEMENT),
    ("task-1", ELEMENT_COMPLETING),
    ("task-1", ELEMENT_COMPLETED),
    ("sequenceFlow_3699d2cf-9108-4f14-b2f7-09b0b74a2952", SEQUENCE_FLOW_TAKEN),
    ("join", ACTIVATE_ELEMENT),
    ("join", ELEMENT_ACTIVATING),
    ("join", ELEMENT_ACTIVATED),
    ("join", ELEMENT_COMPLETING),
    ("join", ELEMENT_COMPLETED),
    ("sequenceFlow_d539266a-c78a-4998-b560-36bfc2514994", SEQUENCE_FLOW_TAKEN),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ACTIVATE_ELEMENT),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATED),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETED),
    ("task-2", COMPLETE_ELEMENT),
    ("task-2", ELEMENT_COMPLETING),
    ("task-2", ELEMENT_COMPLETED),
    ("sequenceFlow_9800ecc9-fc67-4000-989d-abf6d41ade15", SEQUENCE_FLOW_TAKEN),
    ("throw", ACTIVATE_ELEMENT),
    ("throw", ELEMENT_ACTIVATING),
    ("throw", ELEMENT_ACTIVATED),
    ("throw", COMPLETE_ELEMENT),
    ("throw", ELEMENT_COMPLETING),
    ("throw", ELEMENT_COMPLETED),
    ("catch", ACTIVATE_ELEMENT),
    ("catch", ELEMENT_ACTIVATING),
    ("catch", ELEMENT_ACTIVATED),
    ("catch", COMPLETE_ELEMENT),
    ("catch", ELEMENT_COMPLETING),
    ("catch", ELEMENT_COMPLETED),
    ("sequenceFlow_54f9d4b2-0109-40d8-992f-1090b2098f09", SEQUENCE_FLOW_TAKEN),
    ("join", ACTIVATE_ELEMENT),
    ("join", ELEMENT_ACTIVATING),
    ("join", ELEMENT_ACTIVATED),
    ("join", ELEMENT_COMPLETING),
    ("join", ELEMENT_COMPLETED),
    ("sequenceFlow_d539266a-c78a-4998-b560-36bfc2514994", SEQUENCE_FLOW_TAKEN),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ACTIVATE_ELEMENT),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_ACTIVATED),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETING),
    ("endEvent_4537d4bb-cfe7-4d50-8ee2-fa119214d3c8", ELEMENT_COMPLETED),
    ("process", COMPLETE_ELEMENT),
    ("process", ELEMENT_COMPLETING),
    ("process", ELEMENT_COMPLETED)]
to contain only once:
  [("join", ELEMENT_COMPLETED), ("process", ELEMENT_COMPLETED)]
but some elements were found more than once:
  [("join", ELEMENT_COMPLETED)]
```

The inclusive gateway is completed two times. The process instance doesn't wait at the inclusive gateway until the second token arrives.

**Review by saig0 — CHANGES_REQUESTED** [review]
@lzgabel very nice improvements. We're close to finishing the PR. :rocket: 

I found one major issue that the check for link events doesn't work. I added a few minor comments to improve the code. Please have a look. :cookie:

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java [hunk 7]
```diff
@@ -49,6 +49,9 @@ public ProcessInstanceElementActivatingApplier(
   public void applyState(final long elementInstanceKey, final ProcessInstanceRecord value) {
 
     createEventScope(elementInstanceKey, value);
+    final var numberOfTakenSequenceFlows =
+        elementInstanceState.getNumberOfTakenSequenceFlows(
+            value.getFlowScopeKey(), value.getElementIdBuffer());
```
[the remark this answers is not in the document]
Comment (lzgabel) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:54]: ❌ After calling the `cleanupSequenceFlowsTaken` method, we have cleaned up the number of taken sequence flows, so we cannot decrement the active sequence flow correctly.

https://github.com/camunda/camunda/blob/c397bd4d44c23499f8352bf90532922ee39082c8/zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java#L80-L96

**Comment by lzgabel** [comment]
👋 Hi @saig0. Please take a moment to review the latest commit when you have time. 🙇

**Review by saig0 — CHANGES_REQUESTED** [review]
@lzgabel thank you for your fast response. :rocket: 

It looks very good. :+1: I found a few things that don't work in all cases or can be improved. Please have a look. :cake:

**Comment by lzgabel** [comment]
Hi @saig0. Please check this out. 🙇

**Review by saig0 — COMMENTED** [review]
@lzgabel the code looks good to merge. Great! :tada: 

Please rebase the PR. In the meantime, I'll do some manual testing to verify the behavior in more complex scenarios.

**Comment by lzgabel** [comment]
> Please rebase the PR. 

Done.

**Review by saig0 — APPROVED** [review]
@lzgabel LTGM. :partying_face: 

Awesome job on implementing a very complex BPMN element. Thank you for investing all the time. This feature is a great improvement for Camunda. :1st_place_medal: 

---

I created two follow-up issues to fix edge cases:
- https://github.com/camunda/camunda/issues/19460
- https://github.com/camunda/camunda/issues/19464

**Comment by cmur2** [comment]
Merged bypassing branch protections temporarily following this [Slack discussion](https://camunda.slack.com/archives/C071KP5BTHB/p1718702404335999?thread_ts=1718698685.644929&cid=C071KP5BTHB).

**Pull request merged**